libraries/hyperhtml/src/components.js

```diff
@@ -0,0 +1,139 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import 'ce-without-children';
+import 'ce-with-children';
+import 'ce-with-properties';
+import 'ce-with-event';
+
+import {hyper} from 'hyperhtml';
+import * as HyperHTMLELement from 'hyperhtml-element';
+
+export const ComponentWithoutChildren = () => hyper`
+  <div>
+    <ce-without-children id="wc"></ce-without-children>
+  </div>`;
+
+export const ComponentWithChildren = () => hyper`
+  <div>
+    <ce-with-children id="wc"></ce-with-children>
+  </div>`;
+
+export class ComponentWithChildrenRerender extends HyperHTMLELement {
+  get defaultState() { return {count: 1}; }
+  created() {
+    this.render(); // first render with default state.count
+  }
+  connectedCallback() {
+    // increment by one and auto-render
+    this.setState({count: this.state.count + 1});
+  }
+  render() { this.html`
+    <div>
+      <ce-with-children id="wc">${this.state.count}</ce-with-children>
+    </div>`;
+  }
+}
+ComponentWithChildrenRerender.define('with-children');
+
+export class ComponentWithDifferentViews extends HyperHTMLELement {
+  get defaultState() { return {showWC: true}; }
+  created() { this.render(); }
+  toggle() { this.setState((prev) => ({showWC: !prev.showWC})); }
+  render() { this.html`
+    <div>
+    showWC: ${this.state.showWC /* just to show-off */}
+    ${this.state.showWC ?
+      // wire the node with an ID to relate once sub components
+      hyper(this, ':ce')`<ce-with-children id="wc"></ce-with-children>` :
+      hyper(this, ':dummy')`<div id="dummy">Dummy view</div>`
+    }
+    </div>`;
+  }
+}
+ComponentWithDifferentViews.define('with-children-diff-views');
+
+export const ComponentWithProperties = () => hyper`
+  <div>
+    <ce-with-properties id="wc"
+      bool=${true}
+      num=${42}
+      str=${'hyperHTML'}
+      data=${{
+        arr: ['h', 'y', 'p', 'e', 'r', 'H', 'T', 'M', 'L'],
+        obj: {org: 'viperHTML', repo: 'hyperHTML'}
+      }}
+    ></ce-with-properties>
+  </div>`;
+
+export class ComponentWithImperativeEvent extends HyperHTMLELement {
+  created() {
+    this.eventClicks = 0;
+    this.eventHandled = false;
+    this.render();
+  }
+  handleCamel() {
+    this.eventClicks++;
+    this.eventHandled = true;
+    this.render();
+  }
+  render() {
+    (this.html`
+      <div>
+        <div id="handled">${this.eventHandled}</div>
+        <ce-with-event ref="customElement" id="wc"></ce-with-event>
```

Comment:
Is the `ref="customElement"` used anywhere else?

Reply:
copy/paste from here?
copy/paste from here?
https://github.com/webcomponents/custom-elements-everywhere/blob/master/libraries/canjs/src/components.js#L142